### PR TITLE
ztp: Add ztp-deploy-wave annotation in generated policy

### PR DESCRIPTION
**Pull request by Missxiaoguo**
In order to make ClusterGroupUpgrade operator determine the policies
deploy order, 'ran.openshift.io/ztp-deploy-wave' annotation is added
to each policy according to the sourceCRs wave annotation.
See, https://github.com/openshift-kni/cnf-features-deploy/pull/862

Resources have different waves means they need to be applied in order, so
one policy cannot have resources with different waves. PolicyGen will raise
error if it happens.
The wave annotation can be overwritten/added via PGT. 

Jira: https://issues.redhat.com/browse/CNF-3635
Signed-off-by: Angie Wang <angwang@redhat.com>
/cc @imiller0 @serngawy @browsell

### CODE REVIEW

**Review by serngawy — CHANGES_REQUESTED** [review]
this is not needed in PGT , just compute the order in the upgrade operator AND no need to annotate the policy all what u need to is to set the order in the generated UOCR

**Comment by Missxiaoguo** [comment]
> this is not needed in PGT , just compute the order in the upgrade operator AND no need to annotate the policy all what u need to is to set the order in the generated UOCR

We don't allow one policy to have resources with different waves. PGT should validate this when building the policy.  And also adding wave in policy is much easier for upgrade operator to query.

File: ztp/policygenerator/policyGen/policyBuilder.go [hunk 1]
```diff
@@ -320,9 +331,12 @@ func (pbuilder *PolicyBuilder) createAcmPolicy(name string, namespace string, re
 	for idx, resource := range resources {
 		objTemp := BuildObjectTemplate(resource)
 		objTempArr[idx] = objTemp
+
+		if err = SetPolicyDeployWave(acmPolicy.Metadata, len(objTempArr), resource); err != nil {
```
Comment (imiller0) [inline ztp/policygenerator/policyGen/policyBuilder.go:335]: Maybe double-check the use of len(objTempArr) here. On line 329 the array is constructed using the len(resources). I believe this will call SetPolicyDeployWave with the same value for this second arg on each iteration. From the code below it looks like you want to call it with '1' the first iteration and any other value on subsequent iterations. Maybe pass "idx+1"?

File: ztp/policygenerator/policyGen/policyBuilder_test.go [hunk 2]
```diff
@@ -479,3 +480,117 @@ spec:
...
+
+	// Run the PGT through the generator
+	pBuilder = NewPolicyBuilder(fHandler)
+	policies, err = pBuilder.Build(pgt)
+
+	// Validate the run
+	assert.NotNil(t, err)
+	assert.Contains(t, err.Error(), "doesn't match with Policy")
+	assert.NotNil(t, policies)
+}
```
Comment (imiller0) [inline ztp/policygenerator/policyGen/policyBuilder_test.go:596]: Consider adding two more tests

1. If the user has an overlay that changes a wave so that the policy ends up with multiple values
2. If the user has an overlay that "fixes" a mismatched set of waves in the source.

File: ztp/policygenerator/policyGen/policyBuilder.go [hunk 3]
```diff
@@ -201,8 +200,20 @@ func (pbuilder *PolicyBuilder) getCustomResource(sourceFile utils.SourceFile, so
 		resourceMap["metadata"].(map[string]interface{})["labels"] = sourceFile.Metadata.Labels
 	}
 	if len(sourceFile.Metadata.Annotations) != 0 {
-		resourceMap["metadata"].(map[string]interface{})["annotations"] = sourceFile.Metadata.Annotations
+		// covert from map[string]string to map[string]interface{}
```
[the remark this answers is not in the document]
Comment (Missxiaoguo) [reply ztp/policygenerator/policyGen/policyBuilder.go]: Done

File: ztp/policygenerator/policyGen/policyBuilder.go [hunk 4]
```diff
@@ -201,8 +200,20 @@ func (pbuilder *PolicyBuilder) getCustomResource(sourceFile utils.SourceFile, so
...
+		for k, v := range sourceFile.Metadata.Annotations {
+			sourceFileAnnotations[k] = v
+		}
+
+		if resourceMap["metadata"].(map[string]interface{})["annotations"] == nil {
+			// source CR does not have annotations
+			resourceMap["metadata"].(map[string]interface{})["annotations"] = make(map[string]interface{})
+		}
+		resourceMap["metadata"].(map[string]interface{})["annotations"] = pbuilder.setValues(
+			resourceMap["metadata"].(map[string]interface{})["annotations"].(map[string]interface{}), sourceFileAnnotations)
```
[the remark this answers is not in the document]
Comment (Missxiaoguo) [reply ztp/policygenerator/policyGen/policyBuilder.go:225]: I agree. I updated the overlay for labels as well and related unittests.

File: ztp/policygenerator/policyGen/policyHelper.go [hunk 5]
```diff
@@ -80,3 +81,42 @@ func BuildObjectTemplate(resource generatedCR) utils.ObjectTemplates {
...
+				return nil
+			}
+		} else if !foundCrDeployWave {
+			// no wave for policy and resource
+			return nil
+		}
+
+		// resource wave doesn't match with policy wave
+		// resource has no wave but policy has or vice versa
+		return errors.New("'ran.openshift.io/ztp-deploy-wave' annotation in Resource " +
```
[the remark this answers is not in the document]
Comment (Missxiaoguo) [reply ztp/policygenerator/policyGen/policyHelper.go]: Done

File: ztp/policygenerator/policyGen/policyBuilder.go (hunk 1 above)
Comment (Missxiaoguo) [reply ztp/policygenerator/policyGen/policyBuilder.go]: You are obsoultely right! Thanks for finding this out. Updated.

File: ztp/policygenerator/policyGen/policyBuilder_test.go (hunk 2 above)
Comment (Missxiaoguo) [reply ztp/policygenerator/policyGen/policyBuilder_test.go:680]: Added.

File: ztp/policygenerator/policyGen/policyBuilderOverlay_test.go [hunk 6]
```diff
@@ -74,6 +75,19 @@ spec:
...
+	assert.NotNil(t, annotations)
+	assert.Equal(t, len(annotations), 2)
+	assert.Equal(t, annotations["key1"], "value1")
+	assert.Equal(t, annotations["key2"], "value2")
+
+	labels := objDef["metadata"].(map[string]interface{})["labels"].(map[string]interface{})
+	assert.NotNil(t, labels)
+	assert.Equal(t, len(labels), 2)
+	assert.Equal(t, annotations["key1"], "value1")
+	assert.Equal(t, annotations["key2"], "value2")
```
Comment (imiller0) [inline ztp/policygenerator/policyGen/policyBuilderOverlay_test.go:90]: Did you intend to validate the labels here rather than the annotations again?

File: ztp/policygenerator/policyGen/policyBuilderOverlay_test.go [hunk 7]
```diff
@@ -103,6 +117,13 @@ spec:
   sourceFiles:
     - fileName: GenericCR.yaml
       policyName: "gen-policy1"
+      metadata:
+        labels:
+          key1: newvalue
+          key3: value3
```
Comment (imiller0) [inline ztp/policygenerator/policyGen/policyBuilderOverlay_test.go:123]: Use key/values for labels that are different from annotations to catch any bugs where the code accidentally uses the wrong datastructure.

File: ztp/policygenerator/policyGen/policyBuilderOverlay_test.go [hunk 8]
```diff
@@ -310,6 +346,11 @@ spec:
   sourceFiles:
     - fileName: GenericDataCR.yaml
       policyName: "gen-policy1"
+      metadata:
+        labels:
+          key1: value1
```
Comment (imiller0) [inline ztp/policygenerator/policyGen/policyBuilderOverlay_test.go:351]: Same note on unique values.

File: ztp/policygenerator/policyGen/testData/GenericSourceFiles/GenericCR.yaml [hunk 9]
```diff
@@ -4,6 +4,12 @@ kind: JustForTest
 metadata:
   name: test123
   namespace: generic-ns
+  labels:
+    key1: value1
+    key2: value2
```
Comment (imiller0) [inline ztp/policygenerator/policyGen/testData/GenericSourceFiles/GenericCR.yaml:9]: See note above on unique values.

File: ztp/source-crs/validatorCRs/informDuValidator.yaml [hunk 10]
```diff
@@ -27,10 +27,10 @@ status:
         name: load-sctp-module-master
       - apiVersion: machineconfiguration.openshift.io/v1
         kind: MachineConfig
-        name: 02-master-workload-partitioning-master
+        name: 02-master-workload-partitioning
```
Comment (imiller0) [inline ztp/source-crs/validatorCRs/informDuValidator.yaml:30]: We should factor this out into a separate PR in case it needs to be backported.

File: ztp/policygenerator/policyGen/policyBuilderOverlay_test.go (hunk 6 above)
Comment (Missxiaoguo) [reply ztp/policygenerator/policyGen/policyBuilderOverlay_test.go:90]: Good catch!! Thank you. Will update

File: ztp/policygenerator/policyGen/policyBuilderOverlay_test.go (hunk 7 above)
Comment (Missxiaoguo) [reply ztp/policygenerator/policyGen/policyBuilderOverlay_test.go]: Updated

File: ztp/source-crs/validatorCRs/informDuValidator.yaml (hunk 10 above)
Comment (Missxiaoguo) [reply ztp/source-crs/validatorCRs/informDuValidator.yaml]: Sure. will create a separate PR for this. Removed from this PR.

**Review by imiller0 — APPROVED** [review]
/lgtm
